### PR TITLE
Protobuf conversion

### PR DESCRIPTION
**Pull request by sgerbino**

### CODE REVIEW

File: libraries/chain/include/koinos/chain/system_calls.hpp [hunk 1]
```diff
@@ -21,7 +23,7 @@ namespace koinos::chain {
 class apply_context;
 class thunk_dispatcher;
 
-std::optional< thunk_id > get_default_system_call_entry( system_call_id sid );
+// std::optional< thunk_id > get_default_system_call_entry( system_call_id sid );
```
Comment (sgerbino) [inline libraries/chain/include/koinos/chain/system_calls.hpp:26]: Let's remove this commented code.

File: libraries/chain/include/koinos/chain/thunk_dispatcher.hpp [hunk 2]
```diff
@@ -30,8 +255,8 @@ namespace detail
    typename std::enable_if< std::is_same< ThunkReturn, void >::value, int >::type
    call_thunk_impl( const std::function< ThunkReturn(apply_context&, ThunkArgs...) >& thunk, apply_context& ctx, char* ret_ptr, uint32_t ret_len, ArgStruct& arg )
    {
-      static_assert( std::is_same< RetStruct, void_type >::value, "Thunk return does not match defined return in koinos-types" );
-      auto thunk_args = std::tuple_cat( std::tuple< apply_context& >( ctx ), pack::reflector< ArgStruct >::make_tuple( arg ) );
+      // static_assert( std::is_same< RetStruct, std::void_t >::value, "Thunk return does not match defined return in koinos-types" );
```
Comment (sgerbino) [inline libraries/chain/include/koinos/chain/thunk_dispatcher.hpp:258]: Let's remove this commented code.

File: libraries/chain/include/koinos/chain/thunk_dispatcher.hpp [hunk 3]
```diff
@@ -67,40 +298,41 @@ namespace detail
...
-      void call_thunk( thunk_id id, apply_context& ctx, char* ret_ptr, uint32_t ret_len, const char* arg_ptr, uint32_t arg_len )const;
+      void call_thunk( uint32_t id, apply_context& ctx, char* ret_ptr, uint32_t ret_len, const char* arg_ptr, uint32_t arg_len )const;
 
       template< typename ThunkReturn, typename... ThunkArgs >
-      auto call_thunk( thunk_id id, apply_context& ctx, ThunkArgs&... args ) const
+      auto call_thunk( uint32_t id, apply_context& ctx, ThunkArgs&... args ) const
       {
          auto it = _pass_through_map.find( id );
-         KOINOS_ASSERT( it != _pass_through_map.end(), thunk_not_found, "Thunk ${id} not found", ("id", static_cast< thunk_id >( id ) ) );
+         KOINOS_ASSERT( it != _pass_through_map.end(), thunk_not_found, "Thunk ${id} not found", ("id", id ) );
```
Comment (sgerbino) [inline libraries/chain/include/koinos/chain/thunk_dispatcher.hpp:307]: Message should begin with lower case `thunk`.

File: libraries/chain/include/koinos/chain/thunk_dispatcher.hpp (hunk 2 above)
Comment (mvandeberg) [inline libraries/chain/include/koinos/chain/thunk_dispatcher.hpp:258]: Should this have been commented out? If so, can we remove the comment?

File: libraries/chain/include/koinos/chain/thunk_dispatcher.hpp [hunk 4]
```diff
@@ -41,8 +266,14 @@ namespace detail
...
       static_assert( std::is_same< RetStruct, ThunkReturn >::value, "Thunk return does not match defined return in koinos-types" );
-      auto thunk_args = std::tuple_cat( std::tuple< apply_context& >( ctx ), pack::reflector< ArgStruct >::make_tuple( arg ) );
-      pack::to_c_str< RetStruct >( ret_ptr, ret_len, std::apply( thunk, thunk_args ) );
+      auto thunk_args = std::tuple_cat( std::tuple< apply_context& >( ctx ), message_to_tuple< ThunkArgs... >( arg ) );
+      auto ret = std::apply( thunk, thunk_args );
+      std::string s;
+      ret.SerializeToString( &s );
+      // Maybe Koinos::Exception?
+      assert( s.size() <= ret_len );
+      KOINOS_TODO( "I think this copy can be optimized away possible with a string stream" );
```
Comment (mvandeberg) [inline libraries/chain/include/koinos/chain/thunk_dispatcher.hpp:275]: TODO needs issue

File: libraries/chain/include/koinos/chain/thunk_dispatcher.hpp [hunk 5]
```diff
@@ -41,8 +266,14 @@ namespace detail
    call_thunk_impl( const std::function< ThunkReturn(apply_context&, ThunkArgs...) >& thunk, apply_context& ctx, char* ret_ptr, uint32_t ret_len, ArgStruct& arg )
    {
       static_assert( std::is_same< RetStruct, ThunkReturn >::value, "Thunk return does not match defined return in koinos-types" );
-      auto thunk_args = std::tuple_cat( std::tuple< apply_context& >( ctx ), pack::reflector< ArgStruct >::make_tuple( arg ) );
-      pack::to_c_str< RetStruct >( ret_ptr, ret_len, std::apply( thunk, thunk_args ) );
+      auto thunk_args = std::tuple_cat( std::tuple< apply_context& >( ctx ), message_to_tuple< ThunkArgs... >( arg ) );
+      auto ret = std::apply( thunk, thunk_args );
+      std::string s;
+      ret.SerializeToString( &s );
+      // Maybe Koinos::Exception?
```
Comment (mvandeberg) [inline libraries/chain/include/koinos/chain/thunk_dispatcher.hpp:273]: Let's discuss this before merge

File: libraries/chain/system_calls.cpp [hunk 6]
```diff
@@ -191,29 +155,41 @@ struct block_setter
...
 
-   KOINOS_ASSERT( !context.is_in_user_code(), insufficient_privileges, "Calling privileged thunk from non-privileged code" );
+   KOINOS_ASSERT( !context.is_in_user_code(), insufficient_privileges, "calling privileged thunk from non-privileged code" );
 
    auto setter = block_setter( context, block );
-   block.active_data.unbox();
-
-   const multihash& tx_root = block.active_data->transaction_merkle_root;
-   size_t tx_count = block.transactions.size();
+   KOINOS_TODO( "Should this be a more specific exception?" );
```
Comment (mvandeberg) [inline libraries/chain/system_calls.cpp:168]: We need more of this type of data checking on everything else. If not as part of the conversion, we need to create an issue for it.

**Review by mvandeberg — APPROVED** [review]

**Pull request merged**